nexus/sdk-python.md

```diff
@@ -0,0 +1,330 @@
+            MyNexusService,  # or string name "my-nexus-service",
+            "my-nexus-endpoint-name",
+            schedule_to_close_timeout=timedelta(seconds=30),
+        )
+
+    @workflow.run
+    async def run(self, message: str) -> EchoOutput:
+        return await self.nexus_client.execute_operation(
+            MyNexusService.echo,
+            EchoInput(message),
+        )
+
+```
+
+**[ALTERNATIVES CONSIDERED]** One might consider using a traditional (`typing.Protocol` / `abc.ABC`) interface (as Java does). However, while the caller-side Nexus service client might conform to those signatures, the handler-side service implementation would not (see below, but for at least two reasons: (1) operations are not represented by their start methods but rather by nullary factory functions returning an operation instance, and (2) even if the handler-side signature were that of the start method, it contains an option container from the request that would be extraneous in an interface definition).
+
+# 2. Implementing a Nexus operation
+
+Traditionally, in order to define an RPC/HTTP handler, you specify a function to act as the handler, and you arrange for a service or operation instance with a lifetime equal to that of the server, containing state created at server-start time, to be in-scope in the function body. For Python examples, see [gRPC,](https://github.com/grpc/grpc/blob/4e9357bca1408596663a218c0c608a4c0560a867/examples/python/route_guide/route_guide_server.py#L65) [xmlrpc [stdlib]](https://docs.python.org/3/library/xmlrpc.server.html#simplexmlrpcserver-example), [starlette/FastAPI](https://www.starlette.io/applications/#storing-state-on-the-app-instance), [Django class-based-views](https://docs.djangoproject.com/en/5.1/topics/class-based-views/). So, a Python developer writing an RPC/HTTP handler will expect to write something like
+
+```python
+# This code (from Python gRPC docs) is not part of the proposal
+class MyService:  # instantiated at server-start time
+    def my_handler(self, my_input, context) -> MyResult: ...
+```
+
+However, unlike traditional RPC/HTTP handlers, a Nexus operation represents a long-running execution and must support three additional verbs (`get_info`, `get_result`, `cancel`), and must support `start` returning both synchronously and asynchronously (including making that decision at request-handling time).
+
+## 2a. Fully manual class-based operation definition
+
+The “fully manual” way we will support this in Nexus Python is that a user will define a class for the service, and a class for each operation.
+
+**[ALTERNATIVES CONSIDERED]** None. Giving the user the option to define classes at both levels for maximum implementation flexibility is the natural choice for Python (as it was in Java).
+
+```python
+import nexusrpc.handler
+
+@nexusrpc.handler.service(interface=interface.MyNexusService, name="my-service")  # import-time check that interface was implemented
```

Comment:
Can we use a different name for the decorator besides service? I know being in `handler` resolves the ambiguity, but I still think we should avoid ambiguous names across the two modules of the nexusrpc project. Also, similar to operation below, when we talk about "service" we are talking about the contract, not ambiguously talking about the contract or implementation. Maybe `service_impl` or `service_handler` or something.